lib/spack/spack/build_systems/autotools.py

```diff
@@ -291,49 +291,158 @@ def check(self):
+            # in an iterable of one element.
+            condition = '+{name}'.format(name=name)
+            options = [(name, condition in spec)]
+        else:
+            condition = '{name}={value}'
+            options = [
+                (value, condition.format(name=name, value=value) in spec)
+                for value in self.variants[name].values
+            ]
+
         # For each allowed value in the list of values
-        for value in self.variants[name].values:
-            # Check if the value is active in the current spec
-            condition = '{name}={value}'.format(name=name, value=value)
-            activated = condition in spec
+        for option_name, activated in options:
             # Search for an override in the package for this value
-            override_name = '{0}_or_{1}_{2}'.format(active, inactive, value)
+            override_name = '{0}_or_{1}_{2}'.format(
+                activation_word, deactivation_word, option_name
+            )
             line_generator = getattr(self, override_name, None)
             # If not available use a sensible default
             if line_generator is None:
                 def _default_generator(is_activated):
                     if is_activated:
-                        line = '--{0}-{1}'.format(active, value)
-                        if active_parameters is not None and active_parameters(value):  # NOQA=ignore=E501
-                            line += '={0}'.format(active_parameters(value))
+                        line = '--{0}-{1}'.format(activation_word, option_name)
+                        if activation_value is not None and activation_value(option_name):  # NOQA=ignore=E501
+                            line += '={0}'.format(
+                                activation_value(option_name)
+                            )
                         return line
-                    return '--{0}-{1}'.format(inactive, value)
+                    return '--{0}-{1}'.format(deactivation_word, option_name)
                 line_generator = _default_generator
             args.append(line_generator(activated))
         return args
 
-    def with_or_without(self, name, active_parameters=None):
-        """Inspects the multi-valued variant 'name' and returns the configure
-        arguments that activate / deactivate the selected feature.
+    def with_or_without(self, name, activation_value=None):
+        """Inspects a variant and returns the arguments that activate
+        or deactivate the selected feature(s) for the configure options.
+
+        This function works on all type of variants. For bool-valued variants
+        it will return by default ``--with-{name}`` or ``--without-{name}``.
+        For other kinds of variants it will cycle over the allowed values and
+        return either ``--with-{value}`` or ``--without-{value}``.
+
+        If activation is given it will be used to compute a return parameter
```

Comment:
I think the following would be clearer:

> If activation_value is given, then for each possible value of the variant, the option --with-value=activation_value(value) or  --without-value will be added depending on whether or not variant=value is in the spec

(basically I think "name" is confusing in the way it's used here)